pkg/controller/ttl/ttlcontroller.go

```diff
@@ -0,0 +1,238 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ttl
+
+import (
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	"k8s.io/kubernetes/pkg/client/legacylisters"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/informers"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// ObjectTTLAnnotation is a suggestion for kubelet for how long it can cache
+	// an object (e.g. secret, config map) before fetching from apiserver again.
+	ObjectTTLAnnotation = "node.kubernetes.io/ttl"
+)
+
+type TTLController struct {
+	kubeClient clientset.Interface
+
+	// nodeStore is a local cache of nodes.
+	nodeStore listers.StoreToNodeLister
+
+	// Secrets that need to be synced.
+	queue workqueue.RateLimitingInterface
+
+	// Returns true if all underlying informers are synced.
+	hasSynced func() bool
+
+	// Number of nodes in the cluster.
+	// Should be used via atomic operations on it.
+	nodeCount int32
+}
+
+func NewTTLController(nodeInformer informers.NodeInformer, kubeClient clientset.Interface) *TTLController {
+	ttlc := &TTLController{
+		kubeClient: kubeClient,
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ttlcontroller"),
+	}
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ttlc.addNode,
+		UpdateFunc: ttlc.updateNode,
+		DeleteFunc: ttlc.deleteNode,
+	})
+
+	ttlc.nodeStore = *nodeInformer.Lister()
+	ttlc.hasSynced = nodeInformer.Informer().HasSynced
+
+	return ttlc
+}
+
+func (ttlc *TTLController) Run(stopCh <-chan struct{}) {
+	glog.Infof("Starting TTL controller")
```

Comment:
add `defer utilruntime.HandleCrash()` so a mistake here doesn't crash the process